File: sherpout-server/src/main/java/com/sherpout/server/api/training/dto/TrainingPlanDTO.java
package com.sherpout.server.api.training.dto;

import com.sherpout.server.commons.dto.TranslatedStringDTO;
import com.sherpout.server.commons.validation.TranslatedStringValid;
import jakarta.validation.Valid;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class TrainingPlanDTO {
    @Null
    private Long id;

    @NotNull
    @TranslatedStringValid
    private TranslatedStringDTO name;

    @TranslatedStringValid
    private TranslatedStringDTO description;

    @Valid
    @NotNull
    @NotEmpty
    @Size(min = 1, max = 7)
    private List<TrainingPlanDayDTO> days;

    private List<AssignedTrainingPlanDTO> assignments;
}

File: sherpout_mobile/lib/common/dto/training_plan_dto.dart
import 'package:sherpoutmobile/common/dto/training_plan_day_dto.dart';
import 'package:sherpoutmobile/common/dto/translated_string_dto.dart';

class TrainingPlanDTO {
  int? id;
  TranslatedStringDto? name;
  TranslatedStringDto? description;
  List<TrainingPlanDayDTO> days;

  TrainingPlanDTO({
    this.id,
    this.name,
    this.description,
    List<TrainingPlanDayDTO>? days,
  }) : days = days ?? [];

  factory TrainingPlanDTO.fromJson(Map<String, dynamic> json) {
    return TrainingPlanDTO(
      id: json['id'] != null ? json['id'] as int : null,
      name: json['name'] != null ? TranslatedStringDto.fromJson(json['name']) : null,
      description: json['description'] != null ? TranslatedStringDto.fromJson(json['description']) : null,
      days: (json['days'] as List?)
          ?.map((e) => TrainingPlanDayDTO.fromJson(e))
          .whereType<TrainingPlanDayDTO>()
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      if (name != null) 'name': name!.toJson(),
      if (description != null) 'description': description!.toJson(),
      'days': days.map((e) => e.toJson()).toList(),
    };
  }
}